### ISSUE TITLE
[Core] Can't connect to Ray when repeatedly failing actors 

### ISSUE THREAD

**Issue by spolcyn** [ray-project/ray#26337]
### What happened + What you expected to happen

1. Bug: When there is high actor turnover (e.g., many actors attempting to start/failing), connect to the Ray cluster with the Ray client hangs. Sometimes, the GCS also crashes during the high actor turnover.

2. Expected: Cluster connections are unaffected by actor turnover

3. Logs when GCS crashes:
```
(ServeController pid=75543) INFO 2022-07-06 16:43:35,304 controller 75543 deployment_state.py:1216 - Adding 1 replicas to deployment 'Worker'.
(ServeController pid=75543) WARNING 2022-07-06 16:44:04,286 controller 75543 deployment_state.py:1453 - Deployment 'Worker' has 4 replicas that have taken more than 30s to be scheduled. This may be caused by waiting for the cluster to auto-scale, or waiting for a runtime environment to install. Resources required for each replica: {'CPU': 1}, resources available: {}.
(raylet) [2022-07-06 16:44:34,189 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75780) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:34,213 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75783) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(ServeController pid=75543) WARNING 2022-07-06 16:44:34,324 controller 75543 deployment_state.py:1453 - Deployment 'Worker' has 10 replicas that have taken more than 30s to be scheduled. This may be caused by waiting for the cluster to auto-scale, or waiting for a runtime environment to install. Resources required for each replica: {'CPU': 1}, resources available: {}.
(raylet) [2022-07-06 16:44:34,312 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75784) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:34,317 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75785) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:34,547 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75787) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:34,993 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75795) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:34,998 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75796) have not registered within the timeout. The process is still alive, probably it's hanging during start.
(raylet) [2022-07-06 16:44:35,313 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75797) have not registered within the timeout. The process is still alive, probably it's hanging during start.
2022-07-06 16:44:55,122 WARNING worker.py:1404 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffffc842061ca588f486f82dbfe501000000 Worker ID: 9c7572e9ead0dde8b877723c5961557009ae091f80baaf0f4a21b2db Node ID: 49b377a486453500ce672dc24acd547442e2bf4a6c86245081004ae3 Worker IP address: 127.0.0.1 Worker port: 10109 Worker PID: 75781
2022-07-06 16:44:55,127 WARNING worker.py:1404 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffffc28fff80cb5a44bd59a9c78b01000000 Worker ID: f474dbca40dd9c3e55681d4434fa001fa7cb98d0c2917b49b6c82b01 Node ID: 49b377a486453500ce672dc24acd547442e2bf4a6c86245081004ae3 Worker IP address: 127.0.0.1 Worker port: 10110 Worker PID: 75782
2022-07-06 16:44:55,143 WARNING worker.py:1404 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff915cffb064fb2ce219f63c8201000000 Worker ID: af1b12be5b5a8fa9af272a302e35e8e991e3da9373396d7e3eadbbc7 Node ID: 49b377a486453500ce672dc24acd547442e2bf4a6c86245081004ae3 Worker IP address: 127.0.0.1 Worker port: 10003 Worker PID: 75543
(ServeController pid=75543) [2022-07-06 16:44:55,128 C 75543 28665276] reference_count.cc:434:  Check failed: it->second.submitted_task_ref_count > 0
(ServeController pid=75543) *** StackTrace Information ***
(ServeController pid=75543)     ray::GetCallTrace()
(ServeController pid=75543)     ray::SpdLogMessage::Flush()
(ServeController pid=75543)     ray::SpdLogMessage::~SpdLogMessage()
(ServeController pid=75543)     ray::RayLog::~RayLog()
(ServeController pid=75543)     ray::core::ReferenceCounter::RemoveSubmittedTaskReferences()
(ServeController pid=75543)     ray::core::ReferenceCounter::UpdateSubmittedTaskReferences()
(ServeController pid=75543)     ray::core::TaskManager::OnTaskDependenciesInlined()
(ServeController pid=75543)     std::__1::__function::__func<>::operator()()
(ServeController pid=75543)     ray::core::CoreWorkerMemoryStore::Put()
(ServeController pid=75543)     ray::core::TaskManager::MarkTaskReturnObjectsFailed()
(ServeController pid=75543)     ray::core::TaskManager::FailPendingTask()
(ServeController pid=75543)     ray::core::TaskManager::FailOrRetryPendingTask()
(ServeController pid=75543)     ray::core::CoreWorkerDirectActorTaskSubmitter::DisconnectActor()
(ServeController pid=75543)     ray::core::ActorManager::HandleActorStateNotification()
(ServeController pid=75543)     std::__1::__function::__func<>::operator()()
(ServeController pid=75543)     EventTracker::RecordExecution()
(ServeController pid=75543)     std::__1::__function::__func<>::operator()()
(ServeController pid=75543)     boost::asio::detail::completion_handler<>::do_complete()
(ServeController pid=75543)     boost::asio::detail::scheduler::do_run_one()
(ServeController pid=75543)     boost::asio::detail::scheduler::run()
(ServeController pid=75543)     boost::asio::io_context::run()
(ServeController pid=75543)     ray::core::CoreWorker::RunIOService()
(ServeController pid=75543)     std::__1::__thread_proxy<>()
(ServeController pid=75543)     _pthread_start
(ServeController pid=75543)     thread_start
(ServeController pid=75543)
(Worker pid=75781) [2022-07-06 16:44:55,104 C 75781 28669693] gcs_client.cc:300: Couldn't reconnect to GCS server. The last attempted GCS server address was 127.0.0.1:6379
(Worker pid=75781) *** StackTrace Information ***
(Worker pid=75781)     ray::GetCallTrace()
(Worker pid=75781)     ray::SpdLogMessage::Flush()
(Worker pid=75781)     ray::SpdLogMessage::~SpdLogMessage()
(Worker pid=75781)     ray::RayLog::~RayLog()
(Worker pid=75781)     ray::gcs::GcsClient::ReconnectGcsServer()
(Worker pid=75781)     ray::gcs::GcsClient::PeriodicallyCheckGcsConnection()
(Worker pid=75781)     ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()
(Worker pid=75781)     std::__1::__function::__func<>::operator()()
(Worker pid=75781)     EventTracker::RecordExecution()
(Worker pid=75781)     boost::asio::detail::binder1<>::operator()()
(Worker pid=75781)     boost::asio::detail::wait_handler<>::do_complete()
(Worker pid=75781)     boost::asio::detail::scheduler::do_run_one()
(Worker pid=75781)     boost::asio::detail::scheduler::run()
(Worker pid=75781)     boost::asio::io_context::run()
(Worker pid=75781)     ray::core::CoreWorker::RunIOService()
(Worker pid=75781)     std::__1::__thread_proxy<>()
(Worker pid=75781)     _pthread_start
(Worker pid=75781)     thread_start
(Worker pid=75781)
(Worker pid=75782) [2022-07-06 16:44:55,113 C 75782 28669699] gcs_client.cc:300: Couldn't reconnect to GCS server. The last attempted GCS server address was 127.0.0.1:6379
(Worker pid=75782) *** StackTrace Information ***
(Worker pid=75782)     ray::GetCallTrace()
(Worker pid=75782)     ray::SpdLogMessage::Flush()
(Worker pid=75782)     ray::SpdLogMessage::~SpdLogMessage()
(Worker pid=75782)     ray::RayLog::~RayLog()
(Worker pid=75782)     ray::gcs::GcsClient::ReconnectGcsServer()
(Worker pid=75782)     ray::gcs::GcsClient::PeriodicallyCheckGcsConnection()
(Worker pid=75782)     ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()
(Worker pid=75782)     std::__1::__function::__func<>::operator()()
(Worker pid=75782)     EventTracker::RecordExecution()
(Worker pid=75782)     boost::asio::detail::binder1<>::operator()()
(Worker pid=75782)     boost::asio::detail::wait_handler<>::do_complete()
(Worker pid=75782)     boost::asio::detail::scheduler::do_run_one()
(Worker pid=75782)     boost::asio::detail::scheduler::run()
(Worker pid=75782)     boost::asio::io_context::run()
(Worker pid=75782)     ray::core::CoreWorker::RunIOService()
(Worker pid=75782)     std::__1::__thread_proxy<>()
(Worker pid=75782)     _pthread_start
(Worker pid=75782)     thread_start
(Worker pid=75782)
```

### Versions / Dependencies

Python 3.9.10, Ray 1.13, macOS

### Reproduction script

To repro:
1. Start ray (`ray start --head`)
2. Run script once
3. Wait until `deploy()` has started the actor creation process
4. `Ctrl + C` out of the script
5. Run the script again, connection fails

```
import ray
from ray import serve
from tqdm import tqdm

@serve.deployment(max_concurrent_queries=1)
class Worker():
    def __init__(self):
        raise Exception

ray.init(address="auto")
serve.shutdown()
serve.start(detached=True, http_options={"port": 8081})

Worker.options(version="v1", num_replicas=10).deploy(_blocking=False)

handle = Worker.get_handle()
refs = [handle.do_work.remote(x) for x in tqdm(range(12), desc="Obj refs")]
print(ray.get(refs))
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by rkooo567** [comment]
cc @stephanie-wang is this related to the issue you are fixing now (from the release test)?

**Comment by stephanie-wang** [comment]
The assertion failure looks like it's the same issue as #18908, but I think the GCS connection error is different. Maybe that one is the same as #26299.

**Comment by rkooo567** [comment]
> (raylet) [2022-07-06 16:44:34,317 E 75507 28664986] (raylet) worker_pool.cc:502: Some workers of the worker process(75785) have not registered within the timeout. The process is still alive, probably it's hanging during start.

This is also concerning. Maybe something we should handle before 2.0. (I remember this started happening when we migrated to no redis GCS, so it should be related to internal KV load on worker startup)

**Comment by rkooo567** [comment]
Revisit after https://github.com/ray-project/ray/issues/26299 cc (@iycheng it is highly likely the same cause) for the check failure @stephanie-wang will look at it (same as https://github.com/ray-project/ray/issues/18908).

**Comment by iycheng** [comment]
I can't reproduce it on the master. @spolcyn  can you give it another try?

**Comment by spolcyn** [comment]
@iycheng Looks like it does still fail for me, yep. I noticed this time that it just took a very long time to connect to the cluster (several minutes) and the script hung until that point -- maybe not truly a bug then, just a performance thing on the GCS/client connection that this could be a good test case for?

Good with closing this.

**Comment by spolcyn** [comment]
This works on master (resolved sometime between 1.3 and now), thanks to @iycheng for helping debug
